socorro/unittest/lib/test_libsocorrodataschema.py

```diff
@@ -0,0 +1,740 @@
+            "properties": {
+                "length": {"type": "number"},
+            },
+        }
+        self.validate_schema(schema)
+
+        document = {"length": number}
+        assert self.schema_reduce(schema, document) == document
+
+    def test_invalid_number(self):
+        schema = {
+            "$schema": "moz://mozilla.org/schemas/socorro/socorro-data/1-0-0",
+            "type": "object",
+            "properties": {
+                "years": {"type": "number"},
+            },
+        }
+        self.validate_schema(schema)
+
+        document = {"years": "10"}
+        msg_pattern = r"invalid: .years: type string not in \['number'\]"
+        with pytest.raises(InvalidDocumentError, match=msg_pattern):
+            assert self.schema_reduce(schema, document) == document
+
+    def test_string(self):
+        schema = {
+            "$schema": "moz://mozilla.org/schemas/socorro/socorro-data/1-0-0",
+            "type": "object",
+            "properties": {
+                "name": {"type": "string"},
+            },
+        }
+        self.validate_schema(schema)
+
+        document = {"name": "Joe"}
+        assert self.schema_reduce(schema, document) == document
+
+    def test_object_properties(self):
+        schema = {
+            "$schema": "moz://mozilla.org/schemas/socorro/socorro-data/1-0-0",
+            "type": "object",
+            "properties": {
+                "person": {
+                    "type": "object",
+                    "properties": {
+                        "name": {"type": "string"},
+                        "favorites": {
+                            "type": "object",
+                            "properties": {
+                                "color": {"type": "string"},
+                                "ice_cream": {"type": "string"},
+                            },
+                        },
+                    },
+                },
+            },
+        }
+        self.validate_schema(schema)
+
+        document = {
+            "person": {
+                "name": "Janet",
+                "favorites": {
+                    "color": "blue",
+                    "ice_cream": "chocolate",
+                    "hobby": "napping",
+                },
+            }
+        }
+        expected_document = {
+            "person": {
+                "name": "Janet",
+                "favorites": {
+                    "color": "blue",
+                    "ice_cream": "chocolate",
+                },
+            }
+        }
+        assert self.schema_reduce(schema, document) == expected_document
+
+    def test_object_pattern_properties(self):
+        schema = {
+            "$schema": "moz://mozilla.org/schemas/socorro/socorro-data/1-0-0",
+            "type": "object",
+            "properties": {
+                "registers": {
+                    "type": "object",
+                    "pattern_properties": {
+                        "^r.*$": {"type": "string"},
```

Comment:
Should add a comment here because the `r` is easy to miss. Plus it should be a raw string.